### PR TITLE
[GSOC][TMVA][SOFIE] Max ONNX Operator implemented with the corresponding unit tests

### PR DESCRIPTION
**Pull request by Neel-Shah-29**
# This Pull request:
Adds the Max ONNX Operator and the corresponding unit tests to validate the code.

- [X] tested changes locally

### CODE REVIEW

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+namespace Experimental{
+namespace SOFIE{
+
+template <typename T>
+class ROperator_Max final : public ROperator
+{
+
+private:
+
+   std::string fNX1;
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:20]: The MAX operator can have as input a list of tensors of arbitrary size. Here we are limited to 2, so it is like a binary operator. 
You should have a `std::vector<std::string>` defining  the input tensor names

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      if (model.CheckIfTensorAlreadyExist(fNX1) == false){
+         throw std::runtime_error(std::string("TMVA SOFIE Max Op Input Tensor ") + fNX1 + "is not found in model");
+      }
+      if (model.CheckIfTensorAlreadyExist(fNX2) == false) {
+         throw std::runtime_error(std::string("TMVA SOFIE Max Op Input Tensor ") + fNX2 + "is not found in model");
+      }
+      auto shapeX1 = model.GetTensorShape(fNX1);
+      auto shapeX2 = model.GetTensorShape(fNX2);
+      // assume same shape X1 and X2 
+      if (shapeX1 != shapeX2) {
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:52]: I think we should support broadcasting for the different input shapes, which should be > 2.

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      }
+      std::stringstream out;
+      // int length = 1;
+      // for(auto& i: fShape){
+      //    length *= i;
+      // }
+      size_t length = ConvertShapeToLength(fShape);
+      out << "\n//------ Max\n";
+      out << SP << "for (size_t id = 0; id < " << length << " ; id++){\n";
+      out << SP << SP << "tensor_" << fNY << "[id] = (tensor_" << fNX1 << "[id] > tensor_" << fNX2 << "[id]) ? (tensor_" << fNX1 << "[id]) : (tensor_" << fNX2 << "[id]);\n";
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:75]: The code should take into account if number of inputs is a number different than 2

File: tmva/sofie_parsers/src/RModelParser_ONNX.cxx [hunk 4]
```diff
@@ -159,6 +159,43 @@ std::unique_ptr<ROperator> make_ROperator_Transpose(const onnx::NodeProto& nodep
    return op;
 }
 
+
+std::unique_ptr<ROperator> make_ROperator_Max(const onnx::NodeProto& nodeproto, const onnx::GraphProto& /*graphproto */, std::unordered_map<std::string, ETensorType>& tensor_type){
+
+   ETensorType input_type = ETensorType::UNDEFINED;
+
```
Comment (lmoneta) [inline tmva/sofie_parsers/src/RModelParser_ONNX.cxx:166]: Here it should not be 2 but use 
`nodeproto.input_size()` to retrieve all the input names

File: tmva/sofie_parsers/src/RModelParser_ONNX.cxx [hunk 5]
```diff
@@ -159,6 +159,43 @@ std::unique_ptr<ROperator> make_ROperator_Transpose(const onnx::NodeProto& nodep
...
+      } else {
+         throw std::runtime_error("TMVA::SOFIE ONNX Parser Max op has input tensor" + input_name + " but its type is not yet registered");
+      }
+   }
+
+   std::unique_ptr<ROperator> op;
+
+   switch(input_type){
+   case ETensorType::FLOAT:
+      op.reset(new ROperator_Max<float>(nodeproto.input(0), nodeproto.input(1), nodeproto.output(0)));
```
Comment (lmoneta) [inline tmva/sofie_parsers/src/RModelParser_ONNX.cxx:184]: You should change the signature for constructor ROPerator_Max (see comment above)

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+private:
+
+   std::string fNX1;
+   std::string fNX2;
+   std::string fNY;
+   std::vector<size_t> fShape;
+
+public:
+   ROperator_Max(){}
+   ROperator_Max(std::string nameX1, std::string nameX2, std::string nameY):
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:27]: The signature of the constructor for `ROperator_Max` should be 
`ROperator_Max( const std::vector<std::string> & inputNames, std::string nameY)`

**Review by lmoneta — CHANGES_REQUESTED** [review]
Thank you @Neel-Shah-29 for the PR.
We need to add support for multiple inputs, as specified in the Max operator
https://github.com/onnx/onnx/blob/main/docs/Operators.md#Max

**Comment by Neel-Shah-29** [comment]
Hello @lmoneta , I have made all the required changes and tested locally , it works fine and all the tests passes. Thanks!

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+   std::string Generate(std::string OpName){
+      OpName = "op_" + OpName;
+      if (fShape.empty()) {
+         throw std::runtime_error("TMVA SOFIE Max called to Generate without being initialized first");
+      }
+      std::stringstream out;
+      size_t length = ConvertShapeToLength(fShape);
+      out << "\n//------ Max\n";
+      out << SP << "for (size_t id = 0; id < " << length << " ; id++){\n";
+      out << SP << SP <<"tensor_" << fNY << "[id] = std::max({";
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:76]: It is better here two separate these 3 cases when generating the code: 
- fInputNames.size() == 1 :  `output[id] =  input[0][id]`
- fInputNames.size() == 2 : `output[id] = std::max(input[0][id],input[1][id])`
-  fInputNames.size() > 2 : code as it is now 

By doing this the performances of Max will be optimised for the first two cases.

Comment (Neel-Shah-29) [reply tmva/sofie/inc/TMVA/ROperator_Max.hxx]: Done with the changes.

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/default with flags -Dtmva-sofie=On

File: tmva/sofie/inc/TMVA/ROperator_Max.hxx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+      for (auto &it : fInputNames) {
+         if (model.CheckIfTensorAlreadyExist(it) == false) {
+            throw std::runtime_error("TMVA SOFIE Concat Op Input Tensor " + it + " is not found in model");
+         }
+         fShapeInputs.push_back(model.GetTensorShape(it));
+      }
+      for(size_t i=0; i < fShapeInputs.size()-1; i++){
+         if(fShapeInputs[i] != fShapeInputs[i+1]){
+            std::string msg = "TMVA SOFIE Max Op: Support only inputs with same shape";
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/ROperator_Max.hxx:57]: Here we support only inputs to Max with the same shape. 
we should support Broadcasting, at least to tensor with same length as we are doing for the Binary operators

Comment (Neel-Shah-29) [reply tmva/sofie/inc/TMVA/ROperator_Max.hxx]: Done with the changes and tested locally

**Review by lmoneta — APPROVED** [review]
LGTM !
Thank you for updating the PR for the broadcasting for same length

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/default with flags -Dtmva-sofie=On

**Pull request merged**